library/src/blas1/rocblas_swap.cpp

```diff
@@ -63,23 +48,8 @@ namespace
 
         RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle);
 
-        // Quick return if possible.
-        if(n <= 0)
-            return rocblas_status_success;
-
-        hipStream_t rocblas_stream = handle->rocblas_stream;
-        int         blocks         = (n - 1) / NB + 1;
-        dim3        grid(blocks);
-        dim3        threads(NB);
-
-        if(incx < 0)
-            x -= ptrdiff_t(incx) * (n - 1);
-        if(incy < 0)
-            y -= ptrdiff_t(incy) * (n - 1);
-
-        hipLaunchKernelGGL(swap_kernel, grid, threads, 0, rocblas_stream, n, x, incx, y, incy);
-
-        return rocblas_status_success;
+        constexpr rocblas_int NB = 256;
```

Comment:
Maybe we should be consistent about where we are declaring this NB variable, scal send this as a template parameter to the *impl function

Reply:
I was looking at trtri when I made the change for scal, that's how it does it there. Not sure if there's a standard place to do this outside of that.

Reply:
Instead of hard-coding `constexpr` values inside of functions or files, they can be added as template parameters. Note that if they are added, they should be added at the _beginning_, rather than the _end_, of the template parameter list, so that any _type_ template arguments can still be deduced automatically.

Originally some of these `NB` and similar blocksizes were defined as macros. In my first refactoring, I changed the `#define`s to `constexpr ... =  ...;`. Maybe we should go one step further, and make them template parameters.

Reply:
If NB will be different per data type I guess moving it makes sense, otherwise you are duplicating in for each data type...  not sure why NB is used for number of threads either  :-)

Reply:
In the case of `TRTRI`, there was a need for the blocksizes to change, not just based on type, but based on the calling routine.

There's not much redundancy in:
```
template<rocblas_int NB, typename T>
rocblas_status func(T* a)
{
}

func<256>(a);
```
instead of:
```
template<typename T>
rocblas_status func(T* a)
{
   static constexpr rocblas_int NB = 256;
}

func(a);
```
The lines of code and compiler optimization efficiency is the same. If you need to change the blocksize, you can change it at the call, and it can be different for each call.

Reply:
standardized location as _impl parameter